### PR TITLE
perf(core): Refactor NgZone, decrease size by 1.2Kb

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

File: packages/core/src/application_ref.ts [hunk 1]
```diff
@@ -251,6 +252,22 @@ function _callAndReportToErrorHandler(errorHandler: ErrorHandler, callback: () =
 }
 
 /**
+ * Provides additional options to the boostraping process.
+ *
+ * @stable
+ */
+export interface BoostrapOptions {
```
Comment (vicb) [inline packages/core/src/application_ref.ts]: Seems like this interface is not in the .d.ts (while it is used).

"boostraping" needs a t

File: packages/core/src/application_ref.ts [hunk 2]
```diff
@@ -251,6 +252,22 @@ function _callAndReportToErrorHandler(errorHandler: ErrorHandler, callback: () =
...
+  /**
+   * Optionally specify if which `NgZone` should be used.
```
Comment (vicb) [inline packages/core/src/application_ref.ts]: typo "if which"

File: packages/core/src/application_ref.ts [hunk 3]
```diff
@@ -251,6 +252,22 @@ function _callAndReportToErrorHandler(errorHandler: ErrorHandler, callback: () =
...
+   *
+   * - Provide your own `NgZone` instance.
+   * - `true` - Use default `NgZone` which requires `Zone.js`.
+   * - `false` - Use `NoNgZone` which does nothing.
+   */
+  ngZone?: NgZone|boolean;
```
Comment (vicb) [inline packages/core/src/application_ref.ts]: what about `ngZone?: NgZone|'default'|'none';`

File: packages/core/src/zone/ng_zone.ts [hunk 4]
```diff
@@ -81,37 +81,58 @@ import {EventEmitter} from '../event_emitter';
...
+   */
+  readonly isStable: boolean = true;
+
+  /**
+   * Notifies when code enters Angular Zone. This gets fired first on VM Turn.
+   */
+  readonly onUnstable: EventEmitter<any> = new EventEmitter(false);
+
+  /**
+   * Notifies when there is no more microtasks enqueue in the current VM Turn.
```
Comment (vicb) [inline packages/core/src/zone/ng_zone.ts]: typo "enqueue"+d

File: packages/core/src/zone/ng_zone.ts [hunk 5]
```diff
@@ -81,37 +81,58 @@ import {EventEmitter} from '../event_emitter';
...
-  private _isStable = true;
-  private _nesting: number = 0;
-  private _onUnstable: EventEmitter<any> = new EventEmitter(false);
-  private _onMicrotaskEmpty: EventEmitter<any> = new EventEmitter(false);
-  private _onStable: EventEmitter<any> = new EventEmitter(false);
-  private _onErrorEvents: EventEmitter<any> = new EventEmitter(false);
+  /**
+   * Notifies when the last `onMicrotaskEmpty` has run and there are no more microtasks, which
+   * implies we are about to relinquish VM turn.
+   * This event gets called just once.
```
Comment (vicb) [inline packages/core/src/zone/ng_zone.ts]: > This event gets called just once.

What if the subscriber enqueue microtasks ? (ie should we add a warning not to do it or will they not be executed anyway which would be weird).

Also I would relinquish -> end. Simpler is better unless the nuance brings something.

File: packages/core/src/zone/ng_zone.ts [hunk 6]
```diff
@@ -81,37 +81,58 @@ import {EventEmitter} from '../event_emitter';
...
+   */
+  readonly onStable: EventEmitter<any> = new EventEmitter(false);
+
+  /**
+   * Notify that an error has been delivered.
```
Comment (vicb) [inline packages/core/src/zone/ng_zone.ts]: "Notifies"

Not sure to get "has been delivered" ?

File: packages/core/testing/src/ng_zone_mock.ts [hunk 7]
```diff
@@ -14,11 +14,12 @@ import {EventEmitter, Injectable, NgZone} from '@angular/core';
...
 export class MockNgZone extends NgZone {
-  private _mockOnStable: EventEmitter<any> = new EventEmitter(false);
+  onStable: EventEmitter<any>;
 
-  constructor() { super({enableLongStackTrace: false}); }
-
-  get onStable() { return this._mockOnStable; }
+  constructor() {
+    super({enableLongStackTrace: false});
+    this.onStable = new EventEmitter(false);
```
Comment (vicb) [inline packages/core/testing/src/ng_zone_mock.ts]: any reason not to use a prop initializer here ?

File: packages/core/src/application_ref.ts [hunk 8]
```diff
@@ -209,8 +210,8 @@ export abstract class PlatformRef {
    * @stable
    */
   abstract bootstrapModule<M>(
-      moduleType: Type<M>,
-      compilerOptions?: CompilerOptions|CompilerOptions[]): Promise<NgModuleRef<M>>;
+      moduleType: Type<M>, compilerOptions?: (CompilerOptions&BoostrapOptions)|
```
Comment (vicb) [inline packages/core/src/application_ref.ts]: Why `&` instead of an extra arg ?

File: packages/core/src/application_ref.ts [hunk 9]
```diff
@@ -277,17 +294,17 @@ export class PlatformRef_ extends PlatformRef {
...
-  private _bootstrapModuleFactoryWithZone<M>(moduleFactory: NgModuleFactory<M>, ngZone?: NgZone):
+  bootstrapModuleFactory<M>(moduleFactory: NgModuleFactory<M>, options?: BoostrapOptions):
       Promise<NgModuleRef<M>> {
     // Note: We need to create the NgZone _before_ we instantiate the module,
     // as instantiating the module creates some providers eagerly.
     // So we create a mini parent injector that just contains the new NgZone and
     // pass that as parent to the NgModuleFactory.
-    if (!ngZone) ngZone = new NgZone({enableLongStackTrace: isDevMode()});
+    const ngZoneOption = options && options.ngZone;
+    const ngZone: NgZone = ngZoneOption == false ?
```
Comment (vicb) [inline packages/core/src/application_ref.ts]: use `===`

Comment (vicb) [inline packages/core/src/application_ref.ts]: -1 on nested ternary. Extract to a function ?

File: packages/core/src/application_ref.ts [hunk 10]
```diff
@@ -342,6 +354,19 @@ export class PlatformRef_ extends PlatformRef {
   }
 }
 
+function optionsReducer<T extends Object>(dst: any, objs: T | T[]): T {
+  if (Array.isArray(objs)) {
+    objs.reduce(optionsReducer, dst);
+  } else {
+    for (let key in objs) {
```
Comment (vicb) [inline packages/core/src/application_ref.ts]: `dst = {...dst, ...objs};` ?

File: packages/core/src/zone/ng_zone.ts [hunk 11]
```diff
@@ -273,3 +273,25 @@ function onLeave(zone: NgZonePrivate) {
   zone._nesting--;
   checkStable(zone);
 }
+
+/**
+ * Provides a noop implementation of `NgZone` which does nothing. This zone requires explicit calls
+ * to framework to perform rendering.
```
Comment (vicb) [inline packages/core/src/zone/ng_zone.ts]: > This zone requires explicit calls to framework to perform rendering.

Do we want to expand a little bit on that, ie give one example where you would need to "call the framework" and what this means ?

File: packages/core/src/zone/ng_zone.ts [hunk 12]
```diff
@@ -273,3 +273,25 @@ function onLeave(zone: NgZonePrivate) {
...
+ *
+ * @stable
+ */
+export class NoNgZone implements NgZone {
+  hasPendingMicrotasks: boolean = false;
```
Comment (vicb) [inline packages/core/src/zone/ng_zone.ts]: should 
- properties be readonly ?
- `isStable` be initialized ?

**Comment by vicb** [comment]
First commit: I find the `NgZone` / `NgZonePrivate` stuff weird. What is the reason for not making `NgZone` and interface instead ? breaking change ?

Second commit: deserves a better API docs IMO.

I left comments on both the commits

**Comment by mhevery** [comment]
@juliemr could you look this over and let me know what do you thing about this change and protractor?

**Comment by juliemr** [comment]
I don't see any red flags for Protractor, it doesn't seem like anything will change from the Testability point of view.

**Comment by mhevery** [comment]
Closed in favor of https://github.com/angular/angular/pull/17672

**Pull request closed without merging**